Build failure with `--enable-g=all` and `pmix.h`

@zhenggb72 has reported that when building with PMIX and `--enable-g=all`, the compilation fails with the following error:

```
/opt/ohpc/admin/pmix/include/pmix_common.h: In function 'pmix_value_destruct':
/opt/ohpc/admin/pmix/include/pmix_common.h:2074:34: warning: character constant too long for its type
             free((m)->data.string);
                                  ^
In file included from ./src/include/mpiimpl.h:183:0,
                 from src/mpi/attr/attr_delete.c:8:
./src/include/mpir_mem.h:92:27: warning: statement with no effect [-Wunused-value]
 #define free(a)           'Error use MPL_free'   :::
                           ^
```

After looking at the code of src/mpl/include/mpl_trmem.h, I think the problem is due to the `#define` of the `free()` function when memory tracing is enabled. What happens is `pmix.h` includes `pmix_common.h` which has macros that calls `free()`. In this case, these invocations will be incorrectly replaced by our `#define` and cause the build to fail.

--- a/src/include/mpiimpl.h
+++ b/src/include/mpiimpl.h
@@ -102,6 +102,18 @@ int vsnprintf(char *str, size_t size, const char *format, va_list ap);
 #define __func__ "__func__"
 #endif
 
+#ifdef USE_PMIX_API
+/* pmix.h contains inline functions that calls malloc, calloc, and free,
+   and it will break with MPL's memory tracing when enabled.
+   Make sure it is included *before* mpl.h.
+*/
+#include <pmix.h>
+#elif defined(USE_PMI2_API)
+#include <pmi2.h>
+#else
+#include <pmi.h>
+#endif
+
 /*****************************************************************************
  * We use the following ordering of information in this file:
  *

--- a/src/include/mpir_process.h
+++ b/src/include/mpir_process.h
@@ -12,10 +12,6 @@
 #include "hwloc.h"
 #endif
 
-#ifdef USE_PMIX_API
-#include "pmix.h"
-#endif
-
 #ifdef HAVE_NETLOC
 #include "netloc_util.h"
 #endif

--- a/src/mpid/ch4/include/mpidimpl.h
+++ b/src/mpid/ch4/include/mpidimpl.h
@@ -18,14 +18,6 @@
 #include <assert.h>
 #endif
 
-#ifdef USE_PMIX_API
-#include <pmix.h>
-#elif defined(USE_PMI2_API)
-#include <pmi2.h>
-#else
-#include <pmi.h>
-#endif
-
 #define MPICH_SKIP_MPICXX
 #include "mpiimpl.h"
 

--- a/src/mpid/ch4/shm/src/shm_impl.c
+++ b/src/mpid/ch4/shm/src/shm_impl.c
@@ -4,11 +4,8 @@
  *      See COPYRIGHT in top-level directory.
  */
 
-#include "mpl.h"
-
-MPL_SUPPRESS_OSX_HAS_NO_SYMBOLS_WARNING;
-
 #include "posix_impl.h"
+MPL_SUPPRESS_OSX_HAS_NO_SYMBOLS_WARNING;
 
 #ifndef SHM_INLINE
 #ifndef SHM_DISABLE_INLINES

--- a/src/mpid/common/bc/mpidu_bc.c
+++ b/src/mpid/common/bc/mpidu_bc.c
@@ -7,7 +7,6 @@
 
 #include "mpidimpl.h"
 #include "mpidu_shm.h"
-#include "mpl.h"
 #include "mpidu_bc.h"
 
 static MPIDU_shm_seg_t memory;

--- a/src/mpid/common/bc/mpidu_bc.c
+++ b/src/mpid/common/bc/mpidu_bc.c
@@ -83,7 +83,6 @@ int MPIDU_bc_allgather(MPIR_Comm * comm, int *nodemap, void *bc, int bc_len, int
 }
 
 #ifdef USE_PMIX_API
-#include <pmix.h>
 
 #define VALLEN 1024
 #define KEYLEN 64
@@ -201,7 +200,6 @@ int MPIDU_bc_table_create(int rank, int size, int *nodemap, void *bc, int bc_len
 }
 
 #elif defined(USE_PMI2_API)
-#include <pmi2.h>
 
 int MPIDU_bc_table_create(int rank, int size, int *nodemap, void *bc, int bc_len, int same_len,
                           int roots_only, void **bc_table, size_t ** bc_indices)

--- a/src/mpid/common/shm/mpidu_shm_alloc.c
+++ b/src/mpid/common/shm/mpidu_shm_alloc.c
@@ -6,14 +6,6 @@
 #include <mpidimpl.h>
 #include "mpl_shm.h"
 
-#ifdef USE_PMI2_API
-#include "pmi2.h"
-#elif defined(USE_PMIX_API)
-#include "pmix.h"
-#else
-#include "pmi.h"
-#endif
-
 #include <stdlib.h>
 #ifdef HAVE_UNISTD_H
 #include <unistd.h>

--- a/src/mpl/include/mpl_trmem.h
+++ b/src/mpl/include/mpl_trmem.h
@@ -7,6 +7,14 @@
 #ifndef MPL_TRMEM_H_INCLUDED
 #define MPL_TRMEM_H_INCLUDED
 
+/* Sometime we have memory allocated from external library but requires
+ * us to free. Use MPL_external_free for these cases.
+*/
+MPL_STATIC_INLINE_PREFIX void MPL_external_free(void *buf)
+{
+    free(buf);
+}
+
 #if defined MPL_NEEDS_STRDUP_DECL && !defined strdup
 extern char *strdup(const char *);
 #endif /* MPL_NEEDS_STRDUP_DECL */

--- a/src/nameserv/pmi/pmi_nameserv.c
+++ b/src/nameserv/pmi/pmi_nameserv.c
@@ -10,13 +10,6 @@
 
 #include "mpiimpl.h"
 #include "namepub.h"
-#ifdef USE_PMI2_API
-#include "pmi2.h"
-#elif defined(USE_PMIX_API)
-#include "pmix.h"
-#else
-#include "pmi.h"
-#endif
 
 /* style: allow:fprintf:1 sig:0 */
 /* For writing the name/service pair */

--- a/src/util/nodemap/build_nodemap.h
+++ b/src/util/nodemap/build_nodemap.h
@@ -435,7 +435,8 @@ static inline int MPIR_NODEMAP_build_nodemap(int sz,
             node = strtok(NULL, ",");
         }
         *out_max_node_id = node_id - 1;
-        MPL_free(nodelist);
+        /* PMIx latest adds pmix_free. We should switch to that at some point */
+        MPL_external_free(nodelist);
         PMIX_PROC_FREE(procs, nprocs);
     }
 #else /* USE_PMI2_API */

--- a/src/util/nodemap/build_nodemap.h
+++ b/src/util/nodemap/build_nodemap.h
@@ -8,13 +8,6 @@
 #define BUILD_NODEMAP_H_INCLUDED
 
 #include "mpl.h"
-#ifdef USE_PMIX_API
-#include "pmix.h"
-#elif defined(USE_PMI2_API)
-#include "pmi2.h"
-#else
-#include "pmi.h"
-#endif
 
 /*
 === BEGIN_MPI_T_CVAR_INFO_BLOCK ===

